pypeit/edgetrace.py

```diff
@@ -0,0 +1,4122 @@
+                       '        [[slitedges]]' + msgs.newline() +
+                       '            sync_predict = nearest')
+#            self.par['sync_predict'] = 'nearest'
+
+            # NOTE: If the PCA decomposition is possible, the
+            # subsequent call to trace.peak_trace (called by
+            # peak_refine) removes all the existing traces and replaces
+            # them with traces at the peak locations. Those traces are
+            # then fit, regardless of the length of the centroid
+            # measurements. So coming out of peak_refine there are no
+            # traces that are fully masked, which is not true if that
+            # block of code isn't run. That means for the left-right
+            # synchronization to work correctly, we have to remove
+            # fully masked traces. This is done inside sync().
+
+        # Left-right synchronize the traces
+        self.sync()
+        if show_stages:
+            self.show(thin=10, include_img=True, idlabel=True)
+
+        # Add user traces
+        if self.par['add_slits'] is not None:
+            self.add_user_traces(trace.parse_user_slits(self.par['add_slits'], self.det))
+
+        # Remove user traces
+        if self.par['rm_slits'] is not None:
+            self.rm_user_traces(trace.parse_user_slits(self.par['rm_slits'], self.det, rm=True))
+
+        # TODO: Add a parameter and an if statement that will allow for
+        # this.
+        # `peak_refine` ends with the traces being described by a
+        # polynomial. Instead finish by reconstructing the trace models
+        # using the PCA decomposition
+#        self.pca_refine(debug=debug)
+#        if show_stages:
+#            self.show(thin=10, include_img=True, idlabel=True)
+            
+        # TODO: Add mask_refine() when it's ready
+
+        # Add this to the log
+        self.log += [inspect.stack()[0][3]]
+        if save:
+            # Save the object to a file
+            self.save()
+
+    def initial_trace(self, img, bpm=None, det=1, binning='1,1', save=False):
+        r"""
+        Initialize the object for tracing a new image.
+
+        This effectively reinstantiates the object and must be the
+        first method called for tracing an image.  The algorithm:
+            - Lightly boxcar smooths the trace image spectrally.
+            - Replaces bad pixel columns, if a mask is provided.
+            - Applies a Sobel filter to the trace image along columns
+              to detect slit edges using steep positive gradients
+              (left edges) and steep negative gradients (right
+              edges). See
+              :func:`pypeit.core.trace.detect_slit_edges`.
+            - Follows the detected left and right edges along
+              spectrally adjacent pixels to identify coherent traces.
+              See :func:`pypeit.core.trace.identify_traces`.
+            - Performs basic handling of orphaned left or right
```

Comment:
It isn't at the moment, but I kept it around for testing because I wasn't sure that there was some corner case that I was messing up by skipping these steps.  If you're confident we don't need that functionality, I'm fine to remove it.